---
title: "Using the MicroBioMap package"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Using the MicroBioMap package}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## Overview

The human microbiome, particularly in the large intestine, is gaining attention
for its role in health and disease, including conditions like colorectal cancer
and inflammatory bowel disease. Various factors, including genetics and
ethnicity, influence the human microbiome, often tied to geographic regions. As
a result, individuals from different global regions tend to have distinct gut
microbiomes. While research has shown significant variation in the gut
microbiome across countries and social groups, much of the microbiome literature
has focused on a limited range of subjects, raising questions about the
generalizability of these findings.

To address these limitations, researchers have turned to publicly available
microbiome data, revealing an overrepresentation of high-income countries like
the United States and Finland in major microbiome research repositories.
However, significant variations from the Western baseline are observed when
other communities are studied. This suggests that the current understanding of
the microbiome may be limited due to geographic gaps in research.

To bridge this gap and enhance the field's bioinformatic capabilities, the Human
Microbiome Compendium is introduced. It comprises over 168,000 human gut
microbiome samples from 68 countries, processed using advanced tools and made
accessible in various formats, including the MicroBioMap R package and a
dedicated website. This dataset is used to analyze global patterns in microbiome
composition and identify areas where our knowledge of the human gut microbiome
is lacking.

### Primary data processing

Amplicon processing was conducted according to the following criteria: If the
quantity of forward read files corresponded with the number of reverse read
files, we proceeded with paired-end sequencing analysis. However, in cases of a
mismatch or the absence of reverse reads, the project was treated as
single-ended data. In both scenarios, we employed DADA2 version 1.14.0 for data
processing (B. J. Callahan et al. 2016). We utilized broad parameter settings
intended to maximize the inclusion of samples while excluding low-quality data:
Specifically, we did not apply fixed trimming from either end or impose a
maximum read length. Reads shorter than 20 nucleotides, those containing
ambiguous ("N") base calls, and reads aligning to the phiX genome (typically
present as a control in Illumina sequencing runs) were eliminated. Additionally,
quality-based truncation of reads was disabled. Paired-end reads were merged
with a minimum overlap of 20 bases. In certain instances, the merging process
failed, resulting in minimal or no merging of forward and reverse reads, often
associated with sequencing strategies featuring non-overlapping reads. For
studies where less than 50 percent of forward reads merged successfully, we
opted to discard the reverse reads rather than concatenate them. This decision
aimed to prevent failures in merging due to low-quality calls or discrepancies
between forward and reverse read files. In such cases, the reverse reads were
excluded, and the projects were reprocessed as single-ended data. When the
number of forward reads did not align with the number of reverse reads within a
sample, an attempt was made to employ DADA2 to identify the sequence identifier
field in the FASTQ file, facilitating the alignment of salvageable samples. In
cases where this attempt proved unsuccessful, the reverse reads were removed,
and the data was reprocessed as single-ended. Taxonomic assignment was carried
out by DADA2, utilizing the SILVA database release 138.1 as a reference (Quast
et al. 2013; McLaren and Callahan 2021).

## Getting started

### Installation

```{r eval=FALSE}
library(BiocManager)
BiocManager::install("seandavi/MicroBioMap")
```

## Basic usage

```{r message=FALSE}
library(MicroBioMap)
# this operation requires about 4GB of RAM
cpd <- getCompendium()
```

```{r}
cpd
```

### Sample metadata


The `colData` slot of the `cpd` object contains the information about samples.
We can examine the names of the available metadata:

```{r}
names(colData(cpd))
```

We can look at the samples by available regions:

* Region
    ```{r}
table(colData(cpd)$region)
    ```

* Library strategy
    ```{r}
table(colData(cpd)$library_strategy)
    ```

* Instrument
    ```{r}
table(colData(cpd)$instrument)
    ```

* Country
    ```{r}
head(sort(table(colData(cpd)$geo_loc_name), decreasing = TRUE))
    ```


### Microbe (row) metadata

Each row in `cpd` represents one microbe. The `rowData` contains a data.frame-
like set of metadata with columns `kingdom`, `phylum`, etc. 

```{r}
head(rowData(cpd))
```

## Use cases

### Samples in specific Bioprojects

To select samples that belong to specific Bioprojects, we can filter using
the "project" metadata column in the `colData`.

```{r}
head(unique(colData(cpd)$project))
```

For example, create a subset of data belonging to Bioprojects:

- PRJDB10485
- PRJDB10527
- PRJDB10528
- PRJDB10612

```{r}
project_ids <- c("PRJDB10485", "PRJDB10527", "PRJDB10528", "PRJDB10612")
cpd_sub_by_project_ids <- cpd[, colData(cpd)$project %in% project_ids]
cpd_sub_by_project_ids
```

### Samples from a specific world region

To select a subset of samples from a specific world region, subset by the
"columns" of the dataset. The following code ummarized the regions are 
available and the number of samples in each region.

```{r}
sort(table(colData(cpd)$region), decreasing = TRUE)
```
Now, select only those samples from, for example, "Sub-Saharan Africa".

```{r}
cpd_africa <- cpd[, colData(cpd)$region == "Sub-Saharan Africa"]
cpd_africa
```

### Histogram of total bases per sample

```{r warning=FALSE, message=FALSE}
library(ggplot2)
ggplot(as.data.frame(colData(cpd)), aes(x = log10(total_bases))) +
    geom_histogram()
```

### Samples filtered by presence of microbe

There are over 4000 microbes represented in the compendium. Some are fairly rare
and we may want to subset the compendium to only those with a specific
bug in the sample. Here, we pick an arbitrary examplar to show just the 
mechanics. 

```{r}
shig_cpd_counts = counts(cpd)['Bacteria.Proteobacteria.Gammaproteobacteria.Enterobacterales.Enterobacteriaceae.Escherichia-Shigella',]
```

Examine the distribution of abundance across all samples in the compendium.

```{r}
hist(log10(shig_cpd_counts))
```
Picking an arbitrary threshold (after all, this is only an example), we can
limit samples to those with plenty of our bacteria of interest present. 

```{r}
shig_cpd <- cpd[, log10(shig_cpd_counts)>2]
dim(shig_cpd)
```

## sessionInfo

```{r}
sessionInfo()
```
